lyncs_io/dask_io.py

```diff
@@ -0,0 +1,177 @@
+        array : dask/numpy array
+            The array to be written
+        chunks : tuple
+            shape of chunks to split the array into
+
+        Returns:
+        --------
+        array : dask array
+            A lazy evaluated array to be computed on demand
+        """
+
+        if not isinstance(array, self.dask.array.Array):
+            array = self.dask.array.from_array(array, chunks=chunks)
+
+        header = _build_header_from_dask_array(array)
+        offset = _get_dask_array_header_offset(header)
+
+        return self.dask.array.map_blocks(
+            _write_blockwise_to_npy,
+            array,
+            self.filename,
+            header,
+            array.shape,
+            offset,
+            chunks=array.chunks,
+            dtype=array.dtype,
+        )
+
+
+def _get_dask_array_header_offset(header):
+    stream = BytesIO()
+    numpy.lib.format._write_array_header(stream, header)
+
+    return len(stream.getvalue())
+
+
+def _build_header_from_dask_array(array, order="C"):
+    header_dict = {"shape": array.shape}
+    header_dict["fortran_order"] = bool(order == "F")
+    header_dict["descr"] = numpy.lib.format.dtype_to_descr(array.dtype)
+
+    return header_dict
+
+
+def _write_npy_header(filename, header, mode=0o666, dir_fd=None, **kwargs):
+    """
+    Writing npy header in file in a touch-like manner to ensure race-free
+    writing as described in:
+    https://stackoverflow.com/questions/1158076/implement-touch-using-python
+
+    The first task that touches the file is also responsible to write the header
+    during that time before releasing the resources.
+    """
+    flags = os.O_CREAT | os.O_APPEND
+    with os.fdopen(os.open(filename, flags=flags, mode=mode, dir_fd=dir_fd)) as fptr:
+        os.utime(
+            fptr.fileno() if os.utime in os.supports_fd else filename,
+            dir_fd=None if os.supports_fd else dir_fd,
+            **kwargs,
+        )
+        with open(filename, "wb+") as f:
+            numpy.lib.format._write_array_header(f, header)
+
+
+def _write_blockwise_to_npy(
+    array_block, filename, header, shape, offset, block_info=None
+):
+    """
+    Performs a lazy blockwise write of a dask array to file.
+
+    Parameters
+    ----------
+    array_block: list
+        Block array.
+    filename: str
+        Filename where the result will be stored
+    header: dict
+        Numpy header to be written in the file
+    shape: tuple
+        Shape of the global array
+    offset: int
+        offset in bytes to where the
+        data start in the file.
+    block_info: dict
+        contains relevant information to the blocks
+        and chunks of the array. Determined by dask
+        when the function is used in operations like
+        `dask.array.map_blocks`.
+
+    Returns:
+    --------
+    data : slice of the memmap written to the file
+    """
+
+    block_id = block_info[None]["chunk-location"]
+
+    if not os.path.exists(filename):
```

Comment:
This is perfect if the file did not exist.
But what would happen is we are overwriting it?
Then the header will not be updated in this case.

I would prefer it like before where one process is always in charge of writing the header, and the others, if the file does not exist they just touch it.

Reply:
And supporting the above, would be nice to have also a test that tries to overwrite files, with an array of different shape. Can you please do it for all cases?